jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -936,6 +968,65 @@ protected void runJob(Runnable job)
         job.run();
     }
 
+    /**
+     * <p>Attempts to evict the current thread from the pool if {@link #getThreads()} is greater than {@link #getMinThreads()}.</p>
+     * @return true if the current thread was evicted, false otherwise.
+     */
+    protected boolean evict()
+    {
+        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+        while (true)
+        {
+            // No jobs available, should we shrink?
+            int threads = getThreads();
```

Comment:
It may be minor but looks like there is possibility that thread may read stale `getThreads` value as thread which about to exit may not have decrease this value. That can cause number of threads to alive less than min-threads?

Reply:
I think there could be a problem.

It used to work when we could only shrink a single thread per period as if the thread count was stale, then the last shrink time would have been pushed out.

If we are shrinking 10 threads per period, then 10 threads could all wake up, see threads=minthreads+1 and all decide to shrink.  They can all advance the shrink threshold.  If they collide whilst doing so, the they will loop and get the threads count again.  But there is a chance they might not collide on updating the shrink threshhold, so we might let too many threads exit.

thinking....

Reply:
@lorban @sbordet this is a difficult one to solve, as we have 2 Atomic counters, and any attempt to keep them "atomic" is bound to be racey.

Currently, we make no CaS operation on the number of threads when a worker thread exits.  Instead we just decrement the thread count and optionally the idle count.

If we are shrinking, perhaps we should CaS the thread count decrement with the number of threads used to make the decision to shrink.  If we fail that CaS, then we have to restore the shrink interval (or more precisely deduct a shrink period from whatever value it currently is) and then loop and try again.  If we success the CaS, we have to make sure the runner does not further decrement the thread count when the thread exits.

I can give this ago... but most times when we think we are being clever with two atomics, we are wrong!

Reply:
@lorban @sbordet I've implemented this.  It is not too horrid, but hard to convince myself it is not racy.  But I'm thinking the worst case is we don't shrink when we could.

@magibney I'm struggling to see how this can easily be generalised as the shrink() method needs access to two atomics to CaS on them.

Reply:
ahhh so this test may have just been picking up the fact that `ensureThreads()` corrected for overshooting `minThreads` floor -- so maybe the test was triggering the situation after all!